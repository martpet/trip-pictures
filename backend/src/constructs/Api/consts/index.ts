export * from './authorizationHeader';
export * from './lambdaEdgeViewerEvent';
