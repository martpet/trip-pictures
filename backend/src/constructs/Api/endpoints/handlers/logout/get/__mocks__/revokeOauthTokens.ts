export const revokeOauthTokens = vi
  .fn()
  .mockName('revokeOauthTokens')
  .mockResolvedValue(undefined);
