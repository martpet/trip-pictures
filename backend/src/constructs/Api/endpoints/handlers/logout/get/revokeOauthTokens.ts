import { authPaths } from 'lambda-layer';
import fetch from 'node-fetch';

type RevokeTokensProps = {
  authDomain: string;
  refreshToken: string;
  clientId: string;
};

export const revokeOauthTokens = async ({
  authDomain,
  refreshToken,
  clientId,
}: RevokeTokensProps) => {
  const response = await fetch(`https://${authDomain}${authPaths.revoke}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      token: refreshToken,
      client_id: clientId,
    }),
  });

  if (!response.ok) {
    const responseText = await response.text();
    throw new Error(responseText);
  }
};
