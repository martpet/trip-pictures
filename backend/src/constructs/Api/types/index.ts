export * from './Cookies';
export * from './EnvVars';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
export * from './RouteHeaders';
