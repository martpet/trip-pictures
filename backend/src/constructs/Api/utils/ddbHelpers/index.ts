export * from './ddbBatchGet';
export * from './ddbBatchWrite';
export * from './ddbScan';
