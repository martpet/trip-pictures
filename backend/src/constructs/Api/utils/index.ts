export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
