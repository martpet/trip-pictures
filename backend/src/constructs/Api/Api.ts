import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack, RemovalPolicy } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

import { Auth, Photos, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getCdkEnv } from '~/utils';

import { createDistroBehaviors } from './endpoints/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getCdkEnv();
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'http-api');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const logBucket = new Bucket(this, 'Distro-Log-Bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
      logBucket,
    });

    createRoutes({ scope, api, auth, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      distribution,
      origin,
    });

    new ARecord(this, 'A-Record', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
