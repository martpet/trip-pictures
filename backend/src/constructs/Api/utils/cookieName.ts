import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
