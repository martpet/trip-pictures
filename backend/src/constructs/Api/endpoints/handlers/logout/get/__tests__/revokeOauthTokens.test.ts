import fetch, { Response } from 'node-fetch';

import { itCalls, itRejects, itResolves } from '~/constructs/Api/utils';

import { revokeOauthTokens } from '../revokeOauthTokens';

vi.mock('node-fetch');

const args = [
  {
    authDomain: 'dummyAuthDomain',
    refreshToken: 'dummyRefreshToken',
    clientId: 'dummyClientId',
  },
] as Parameters<typeof revokeOauthTokens>;

beforeEach(() => {
  vi.mocked(fetch).mockResolvedValue({ ok: true } as Response);
});

describe('revokeOauthTokens', () => {
  itCalls(fetch, revokeOauthTokens, args);
  itResolves(revokeOauthTokens, args);

  describe('when "fetch" response is not "ok"', () => {
    beforeEach(() => {
      const response = {
        ok: false,
        text: () => Promise.resolve('dummyFetchResponseText'),
      } as Response;
      vi.mocked(fetch).mockResolvedValue(response);
    });
    itRejects(revokeOauthTokens, args);
  });
});
