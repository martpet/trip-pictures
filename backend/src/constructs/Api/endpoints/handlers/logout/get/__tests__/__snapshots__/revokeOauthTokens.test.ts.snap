// Vitest Snapshot v1

exports[`revokeOauthTokens > calls "fetch" with correct args 1`] = `
[
  [
    "https://dummyAuthDomain/oauth2/revoke",
    {
      "body": URLSearchParams {
        Symbol(query): [
          "token",
          "dummyRefreshToken",
          "client_id",
          "dummyClientId",
        ],
        Symbol(context): null,
      },
      "headers": {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`revokeOauthTokens > resolves with a correct value 1`] = `undefined`;

exports[`revokeOauthTokens > when "fetch" response is not "ok" > rejects with a correct value 1`] = `[Error: dummyFetchResponseText]`;
